python_ta/debug/accumulation_table.py

```diff
@@ -0,0 +1,155 @@
+"""
+Table data structure that prints a nicely formatted table
+for an accumulator loop
+"""
+from __future__ import annotations
+
+import inspect
+import sys
+
+import astroid
+import tabulate
+
+
+def num_whitespaces(start_of_loop: str) -> int:
+    """Number of spaces at the beginning of the accumulation loop"""
+    blank_chars = 0
+    for char in start_of_loop:
+        if char.isspace():
+            blank_chars += 1
+        else:
+            break
+
+    return blank_chars
+
+
+def get_loop_lines(lines: list[str], num_whitespace: int) -> str:
+    """Function that returns the lines from the start to the end
+    of the accumulator loop
+    """
+    endpoint = len(lines)
+    for i in range(len(lines)):
+        if lines[i].strip() != "" and not lines[i][num_whitespace].isspace():
+            endpoint = i
+            break
+
+    return "\n".join(lines[:endpoint])
+
+
+def get_for_node(frame: types.FrameType) -> types.NodeNG:
+    """Return the For node from the frame containing the accumulator loop"""
+    func_string = inspect.cleandoc(inspect.getsource(frame))
+    with_stmt_index = inspect.getlineno(frame) - frame.f_code.co_firstlineno
+    lst_str_lines = func_string.splitlines()
+    lst_from_with_stmt = lst_str_lines[with_stmt_index + 1 :]
+    num_whitespace = num_whitespaces(lst_str_lines[with_stmt_index])
+    loop_lines = get_loop_lines(lst_from_with_stmt, num_whitespace)
+
+    return astroid.parse(loop_lines).body[0]
+
+
+class AccumulationTable:
+    """
+    Class used as a form of print debugging to analyze different loop and
+    accumulation variables during each iteration in a for loop
+
+    Private instance attributes:
+        loop_accumulators: a mapping between the accumulation variables
+            and their values during each iteration
+        loop_var_name: the name of the loop variable
+        loop_var_val: the values of the loop variable during each iteration
+        _loop_lineno: the line number of the for loop
+
+    """
+
+    loop_accumulators: dict[str, list]
+    loop_var_name: str
+    loop_var_val: list
+    _loop_lineno: int
+
+    def __init__(self, accumulation_names: list[str]) -> None:
+        self.loop_accumulators = {accumulator: [] for accumulator in accumulation_names}
+        self.loop_var_name = ""
+        self.loop_var_val = []
+        self._loop_lineno = 0
+
+    def _record_iteration(
+        self, accumulator_values: list, loop_variable_value: Any, frame: types.FrameType
```

Comment:
Okay this function is definitely better, but can be simplified even further. The only argument you need to pass in here is the `frame`. The current accumulator values and loop variable value can be computed from `frame.f_locals`; currently, you're doing this outside of the function (in `_trace_loop`), but really you can always do it in this function.

Note that if you do so, the code for updating the *accumulator values* should be the same for the zero-th and every other iteration; the only part that requires branching is choosing the value to append to `self.loop_var_val`.